### PR TITLE
feat(registry): first pass at code and tests

### PR DESCRIPTION
**Pull request by Eihcir0**
https://github.com/albertocuestacanada/solidity-mentorship/issues/1

### CODE REVIEW

File: contracts/Registry.sol [hunk 1]
```diff
@@ -0,0 +1,16 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.4;
+
+contract Registry {
+    mapping(string => address) public claimedNames;
+
```
Comment (albertocuestacanada) [inline contracts/Registry.sol:6]: Natspec, please, at least a @dev line.

File: contracts/Registry.sol [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    function claimName(string memory _name) public {
+        require(claimedNames[_name] == address(0), "Name already claimed");
+        claimedNames[_name] = msg.sender;
```
Comment (albertocuestacanada) [inline contracts/Registry.sol:9]: At the end of each function that modifies the state, you should emit an event if the final value of the modified variables. Events are vital for frontend integration.

File: test/registry/Registry.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    it("should allow names to be claimed", async function () {
+      name = "booyah";
+      user = this.signers.user1;
+
+      await this.registry.connect(user).claimName(name);
+
+      const newOwner = await this.registry.claimedNames(name);
+      expect(newOwner).to.be.equal(user.address);
+    });
+
```
Comment (albertocuestacanada) [inline test/registry/Registry.ts:42]: You can nest a `describe` block inside `#claimName()`, and in the `beforeEach` block of that nested `describe`, you put the function that you just tested to work: `registry.claimName(name)`.

That way you follow a pattern:
 - You are at state 0
 - Test all the things that should revert
 - Test one thing that should work
 - Nest a describe, put the thing that works in the beforeEach, you are at state 1.

You will end with a kind of state tree, testing all the possible reverts, all the possible states, and all the possible successful functions.

**Review by albertocuestacanada — CHANGES_REQUESTED** [review]
Love it. The simpler is always the better. No place for bugs to hide, and minimal gas cost. Some minimal changes requested (comments, events) and a recommendation on writing better tests.

**Review by albertocuestacanada — CHANGES_REQUESTED** [review]
Love it. The simpler is always the better. No place for bugs to hide, and minimal gas cost. Some minimal changes requested (comments, events) and a recommendation on writing better tests.

File: contracts/Registry.sol [hunk 4]
```diff
@@ -0,0 +1,39 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.4;
+
+/// @title Name Registry
```
Comment (Eihcir0) [inline contracts/Registry.sol:4]: These natspecs seem a little overkill.  In the past I try to make my code as self-documenting as possible and only use comments to help clear up those "wtf" moments.

However, I realize there is a strong use case for comments on smart contracts, I'm sure I will figure out what's important over time and by reviewing smart contracts of others.

**Comment by Eihcir0** [comment]
[contract deployed and verified on Kovan](https://kovan.etherscan.io/address/0x2F1C961F580805F5F1FcB7F6981d2d3674325C0d#code)

File: contracts/Registry.sol (hunk 4 above)
Comment (albertocuestacanada) [reply contracts/Registry.sol:4]: Agree, I usually have a single @dev short paragraph. If I need @param or @return natspec that's actually a sign that there is something wrong with the design. Sometimes I add a @notice when there is something important to say (again, a sign that something is wrong)

Then, I tend to add in-line comments, as minimal code can't always be self-evident, or when something was considered that led to a particular design choice. Also when breaking a longer function into steps.

**Review by albertocuestacanada — APPROVED** [review]
Great job. I like that you went straight and to the point. I think you'll sail through the next couple of issues.

**Pull request merged**